### DISCUSSION TITLE
Why does SmartPedal use a lot more CPU than Chowdhury's BYOD?

### DISCUSSION THREAD

**Discussion by 38github** [keyth72/SmartGuitarPedal#17] [Q&A] [answered]
Chowdhury's BYOD plugin has a GuitarML insert/module that use only 2-3 % CPU on my computer when using the Blues Jr preset. How come SmartPedal use 25%? Is there something Chowdhury's plugin BYOD use that SmartPedal could use to lower the CPU usage?

Does it have to do with that his implementation uses the LSTM version?

**Comment by GuitarML** [comment, marked as answer]
Good question, it uses a completely different neural network model called WaveNet, which uses more CPU (4x - 8x depending on the network size). Jatin's plugin uses the LSTM model, which is the same as NeuralPi/TS-M1N3/Chameleon/EpochAmp.  I'm leaving SmartPedal and SmartAmp with the WaveNet model because I think it has it's own advantages, and for people to experiment with. I also think WaveNet has a slightly more natural sound, as it was originally created specifically for audio applications, where LSTM was developed for any kind of time-series prediction (stock prices, weather, etc.). That being said, I am releasing a new SmartPedal-like plugin that uses LSTM (should be out today or tomorrow), and that will be the new "GuitarML standard" that I'll recommend to people going forward.  It will also have it's own Colab script and input wav file in a utility package.

**Reply by 38github** [reply]
Thank you for the answer! I also think WaveNet sounds better/more accurate. LSTM (at least the old versions) generates high frequencies when the models are pushed with what I think is more gain than they were modeled after. It will be interesting though to hear the new GuitarML standard using LSTM and if it compares to WaveNet.

Will the input file be the same as in SmartPedalCapture? I think it would be great if it is since we can easily redo older recordings we have made and remodel them with any neural network you offer.

**Reply by GuitarML** [reply]
@38github Ok the new plugin is out! It's called Proteus (for the greek god of shapeshifting), if you find any bugs with it let me know, I wanted to get it out fast and maybe hop on the capture hype from ToneX. I also made two new Youtube tutorials, one for an amp, and one for a pedal, capturing the full range of a knob (it's really the same process, just thought people would like to see different devices).

Unfortunately it uses a different capture input file, but I have good reasons for doing that. In all my testing, using generated tones such as sine sweeps and white/brown/pink noise helped the LSTM model, but drastically hurt the WaveNet model. So for this one I include the generated signals at the beginning of the input file, and the loss got much lower than just using guitar/bass. The other difference is that the LSTM has a hidden size of 40, which is double NeuralPi, to allow better accuracy.  It didn't seem to hurt cpu usage to a huge extent to make this increase, at least on a modern laptop. 

Also I should mention that Proteus is not currently open source, but I might change this in the future. There's nothing new under the hood that you can't find in my other plugins, besides using a hidden size of 40 and the graphics.

**Comment by 38github** [comment]
Does SmartPedal process stereo? If so, potential CPU reduction using mono?

I remember Chowdhury mentioning something about this might be why Proteus use more CPU compared to his implementation in BYOD. If so, it would be great to see a CPU reduction in SmartPedal too.

On a side note, I do have a lot of posted issues and discussions. Actually, I have them all except one but that is because I absolutely love this thing and want to see it work like a charm.

**Reply by GuitarML** [reply]
No problem, all good feedback! No, SmartPedal doesn’t process stereo, it processes the first channel and copies that over to the second channel, so no cpu improvement to be found there unfortunately.
